File: javascript/rotor.js
class Rotor {
    constructor(rotor_position) {
        rotor_position = rotor_position.toUpperCase();

        this.N = rotor_position.length;

        for (let i = 0; i < this.N; i++) {
            if (!rotor_position.includes(String.fromCharCode(i+65))) {
                alert("Erreur : la position du rotor devrait contenir " + String.fromCharCode(i+65));
                throw new Error("Erreur : la position du rotor devrait contenir " + String.fromCharCode(i+65));
            }
        }

        this.rotor_position = new Array();

        for (let i = 0; i < this.N; i++) {
            this.rotor_position.push(rotor_position.charCodeAt(i)-65)
        }
    }

    tourner() {
        let nouveau_rotor = new Array();

        for(let i = 0; i < this.N; i++) {
            nouveau_rotor.push((this.rotor_position[i]-1+this.N) % this.N);
        }
        for (let i = 0; i < this.N - 1; i++) {
            this.rotor_position[i] = nouveau_rotor[i+1];
        }
        this.rotor_position[this.N - 1] = nouveau_rotor[0];
    }

    coder(nombre) {
        return this.rotor_position[nombre];
    }

    decoder(nombre) {
        return this.rotor_position.indexOf(nombre);
    }
}

export { Rotor };

File: python/rotor.py
alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"

class Rotor:
   def __init__(self, rotor_position):

      rotor_position = rotor_position.upper()

      for i in range(len(rotor_position)):
         if not alphabet[i] in rotor_position:
            raise Exception("La position du rotor devrait contenir la lettre " + alphabet[i])

      self.rotor_position = []

      for lettre in rotor_position:
         self.rotor_position.append(alphabet.index(lettre))

      self.N = len(rotor_position)

   def show_rotor_position(self):
      return self.rotor_position

   def tourner(self):
      nouveau_rotor = []
      for i in self.rotor_position:
         nouveau_rotor.append( (i-1) % self.N)
      self.rotor_position = nouveau_rotor[1:] + nouveau_rotor[:1]

   def coder_nombre(self, nombre):
      return self.rotor_position[nombre]

   def coder_lettre(self, lettre):
      return self.rotor_position[alphabet.index(lettre.upper())]

   def decoder_nombre(self, nombre):
      return self.rotor_position.index(nombre)

   def decoder_lettre(self, lettre):
      return self.rotor_position.index(alphabet.index(lettre.upper()))